### ISSUE TITLE
使用 CTC loss 训练后测试时发现 pred 为空

### ISSUE THREAD

**Issue by WeldonWangwang** [ZhengkunTian/OpenTransformer#46]
Hi,
我在 conformer_baseline.yaml 的基础上将 `ctc_weight: 0.0` 改为 `ctc_weight: 1.0`， 并且加入 `encoder_output_size: 256`

训练完成后使用 python eval.py -m model.pt 进行测试，发现预测结果全部为空，打印了下 decoder interence 中的 preds 和 logits:

```

    def inference(self, preds, memory, memory_mask=None, cache=None):

        assert preds.dim() == 2
        # dec_output = self.embedding(preds)
        # dec_output, pos = self.pos_encoding.inference(dec_output)
        # mask = get_transformer_decoder_mask(preds)

        # new_caches = []
        # attn_weights = {}
        # for i, block in enumerate(self.blocks):
        #     block_cache = cache[i] if cache is not None else {'slf': None, 'src': None}
        #     dec_output, attn_weight, block_cache = block.inference(dec_output, mask, memory, memory_mask.unsqueeze(1), pos, cache=block_cache)
        #     attn_weights['dec_block_%d' % i] = attn_weight
        #     new_caches.append(block_cache)

        # if self.normalize_before:
        #     dec_output = self.after_norm(dec_output)

        # logits = self.output_layer(dec_output) # logits [batch_size, 1, model_size]
        print(preds)
        logits, attn_weights = self.forward(preds,  memory, memory_mask)
        print(logits)
        print('!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!')

        log_probs = F.log_softmax(logits[:, -1, :], dim=-1) # logits [batch_size, 1, model_size]

        return log_probs, cache, attn_weights
```

测试结果部分截取如下：

```
tensor([[1],
        [1],
        [1],
        [1],
        [1]], device='cuda:0')
tensor([[[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]]], device='cuda:0')
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
tensor([[1, 0],
        [1, 1],
        [1, 2],
        [1, 3],
        [1, 4]], device='cuda:0')
tensor([[[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]]], device='cuda:0')
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
tensor([[1, 1, 1],
        [1, 0, 2],
        [1, 0, 1],
        [1, 0, 4],
        [1, 0, 0]], device='cuda:0')
tensor([[[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]],

        [[-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38],
         [-3.1599e-40, -2.7828e-39, -1.0696e-38,  ...,  1.0283e-38,
           2.8544e-39, -1.0548e-38]]], device='cuda:0')
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

```

输出的预测值全部不对，但是当使用 conformer_baseline.yaml 进行训练时，测试结果正常。

请问下当  `ctc_weight: 1.0` 时，使用 eval.py 测试时需要更改什么参数吗？

谢谢～～

**Comment by ZhengkunTian** [comment]
ctc_weight 应该是 [0,1)范围内的值 等于1的时候 模型等效于CTC模型，解码器没有训练到 没有意义 因此不能使用eval.py中默认的解码器进行直接解码，应该使用CTC解码器
